server/bin/pbench-backup-tarballs.py

```diff
@@ -0,0 +1,381 @@
+#!/usr/bin/env python3
+# -*- mode: python -*-
+
+import os
+import sys
+import sh
+import glob
+import base64
+import signal
+import hashlib
+import shutil
+import configtools
+import subprocess
+from argparse import ArgumentParser
+
+from pbench import init_report_template, report_status, _rename_tb_link, \
+                       PbenchConfig, BadConfig, get_es, get_pbench_logger, quarantine
+from s3backup import S3Config
+
+_NAME_ = "pbench-backup-tarballs"
+
+# The link source and destination for this operation of this script.
+_linksrc = "TO-BACKUP"
+_linkdest = "BACKED-UP"
+
+# Global logger for the module, setup in main()
+_logger = None
+
+def backup_data(s3_obj, config, report, logger):
+
+    # make sure the local backup directory is present
+    try:
+        os.mkdir(config.BACKUP)
+    except FileExistsError:
+        # directory already exists, ignore
+        pass
+    except Exception:
+        logger.exception(
+            "os.mkdir: Unable to create backup destination directory: {}\n".format(config.BACKUP))
+        return
+
+    # make sure the S3 bucket exists
+    try:
+        s3_obj.connector.head_bucket(Bucket='{}'.format(s3_obj.bucket_name))
+    except Exception as e:
+        logger.info(
+            "Bucket: {} does not exist or you have no access. {}\n".format(s3_obj.bucket_name, e))
+        return
+
+    # make sure the quarantine directory is present
+    qdir = config.get('pbench-server', 'pbench-quarantine-dir')
+    if not os.path.isdir(qdir):
+        logger.error(
+            "{}: {} does not exist, or is not a directory\n", _NAME_, qdir)
+        return
+
+    tarlist = glob.iglob('{}/*/{}/*.tar.xz'.format(config.ARCHIVE, _linksrc))
+    nbackup, ns3_success, ns3_fail, nquaran, nerrs = 0, 0, 0, 0, 0
+
+    for tb in sorted(tarlist):
+        # resolve the link
+        tar = os.path.realpath(tb)
+
+        resultname = os.path.basename(tar)
+        controller_path = os.path.dirname(tar)
+        controller = os.path.basename(controller_path)
+        s3_resultname = "{}/{}".format(controller, resultname)
+        backup_controller_path = "{}/{}".format(config.BACKUP, controller)
+
+        # make sure the controller is present in local backup directory
+        try:
+            os.mkdir(backup_controller_path)
+        except FileExistsError:
+            # directory already exists, ignore
+            pass
+        except Exception:
+            logger.exception(
+                "os.mkdir: Unable to create backup destination directory: {}\n".format(backup_controller_path))
+            continue
+
+        archive_md5 = ("{}.md5".format(tar))
+        try:
+            with open(archive_md5) as f:
+                archive_md5_hex_value = f.readline().split(" ")[0]
+        except Exception:
+            # md5 file does not exist
+            quarantine(qdir, logger, tb)
+            nquaran += 1
+            nerrs += 1
+            report.write("Quarantine: {}\n".format(resultname))
```

Comment:
The report should only contain summary information. The log contains all the gory details but we don't duplicate that in the report. If there is no need to write anything in the report in this function (which I'm not quite sure about), we might as well not pass it as an argument to this function: we'll return enough information so that the report can be written in the caller.